.github/workflows/haskell-ci.yml

```diff
@@ -0,0 +1,208 @@
+# Haskell source files.
+jobs:
+  CI:
+    runs-on: ${{ inputs.os }}
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v4
+    - name: Install Cabal and GHC
+      uses: haskell-actions/setup@v2
+      id: setup-haskell
+      with:
+        cabal-version: ${{ inputs.cabal }}
+        ghc-version: ${{ inputs.ghc }}
+    - name: Create sdist
+      if: inputs.sdist
+      run: |
+        if [[ "${{ inputs.sdist }}" == true ]]; then
+          cabal sdist -o sdist
+          mkdir -p unpacked
+          tar -C unpacked -xzf sdist/*.tar.gz
+          cd unpacked/*
+        fi
+        echo "WORK_DIR=$(pwd)" >> $GITHUB_ENV
+    - name: Restore Cabal cache
+      uses: actions/cache/restore@v4
+      id: cache
+      env:
+        # Each `inputs.os` (e.g., `ubuntu-22.04-arm`) uniquely determines a
+        # `runner.arch` (e.g., ARM64), so there is no need to include the latter
+        # as part of the cache key.
+        key: cabal-${{ inputs.os }}-ghc${{ inputs.ghc }}
+      with:
+        path: |
+          ${{ steps.setup-haskell.outputs.cabal-store }}
+          ${{ env.WORK_DIR }}/dist-newstyle
+        # Use `github.ref` in the key for branch-specific caching.
+        #
+        # Include `github.sha` to allow saving a new cache even when there was a
+        # cache hit. See https://github.com/actions/cache/issues/109.
+        #
+        # The `restore-key` allows for restoring the cache from a different branch
+        # (e.g., the default branch) if none exists for the current branch (e.g.,
+        # on the first run on a new branch).
+        key: |
+          ${{ env.key }}-${{ github.ref }}-${{ github.sha }}
+        restore-keys: |
+          ${{ env.key }}-${{ github.ref }}
+          ${{ env.key }}-
+    - name: Configure
+      run: cabal configure ${{ inputs.configure-flags }} ${{ inputs.build-targets }} ${{ inputs.test-targets }}
+      working-directory: ${{ env.WORK_DIR }}
+    # Note [Parallelism]
+    # ------------------
+    #
+    # We split the build into two steps:
+    #
+    # - Build only the dependencies. Build them in parallel at the package level
+    #   (which is Cabal's default, see haskell/cabal#5776).
+    # - Build this project (a single package). Use --ghc-options=-j (in the
+    #   `configure` step) to build local packages' modules in parallel.
+    #
+    # This scheme maximizes the use of parallelism while avoiding
+    # oversubscription. See GHC proposal #540 for additional background. It would
+    # be nice to replace this with `cabal configure --semaphore`, but this only
+    # became available with GHC 9.8. To support older versions, we don't use it
+    # just yet.
+    - name: Build dependencies
+      run: cabal build --only-dependencies ${{ inputs.build-targets }} ${{ inputs.test-targets }}
+      working-directory: ${{ env.WORK_DIR }}
+    - name: Build
+      run: cabal build ${{ inputs.build-targets }}
```

Comment:
It seems like both of these steps ("Build dependencies" and "Build" )will be affected by the `configure-flags`, although I suspect you want `-j` supplied to the dependencies build for breadth parallelism across packages but not `--ghc-options=-j`, and vice-versa for the actual build?

Reply:
> although I suspect you want -j supplied to the dependencies build for breadth parallelism across packages but not --ghc-options=-j, and vice-versa for the actual build?

That's correct, and is outlined in `Note [Parallelism]`.

> It seems like both of these steps ("Build dependencies" and "Build" )will be affected by the configure-flags,

See https://github.com/GaloisInc/parameterized-utils/pull/193#discussion_r2019022194 -- `--ghc-options` only applies to local packages.